### ISSUE TITLE
Multiple issues for really large (over 10k pages) websites

### ISSUE THREAD

**Issue by brnquester** [putyourlightson/craft-blitz#492]
We have been struggling to get Blitz working smoothly with our website, to some degree we have succeed, but the difficulties are all related to the size of our content. We need some help to get it right. 

Note: You might notice that the `queryStringCaching` is set to `1`; however, there is a small whitelisted query list that is working properly and is responsible for a very small number of cached pages, less than 50. 

## Problems
- Clear cache via URL API just throw a timeout when we are full cached, all pages already warmed up
- Accessing `CMS > Utilities > Blitz` just throw time out, it doesn't even load
- Queue job for clearing cache are stuck or take more than 2 hours, even when is just for 1 entry
- Scheduled posts never show up on the main listing, most likely due to the clear cache taking way too long

Number of cacheable entries: ~ 10,000
Cache storage average size when warmed up: 2.4Gb

## Technical background
- Craft CMS v3.7.66
- Blitz plugin v3.12.6
- PHP v8.0.28
- DB MySQL v8.0.23
- PHP Memory Limit 1024M
- Max execution time 300s
- Robust queue job handling on via `systemd` worker

## Relevant configs
- Configs overwritten via `config/blitz.php`
```
        'debug' => false,

        'warmCacheAutomatically' => true,

        "cachingEnabled" => true,

        'cacheDuration' => 0,

        'cacheElementQueries' => false,

        'cacheElements' => true,

        "queueJobTtr" => 3600,

        'cacheWarmerType' => 'putyourlightson\blitz\drivers\warmers\GuzzleWarmer',

        "cacheWarmerSettings" => ["concurrency" => 2],

        'cacheStorageType' => 'putyourlightson\blitz\drivers\storage\FileStorage',

        'cacheStorageSettings' => [
           'folderPath' => '@webroot/cache/blitz',
           'createGzipFiles' => false,
           'countCachedFiles' => true,
        ],

        'clearCacheAutomatically' => true,

        'driverJobPriority' => 100,

        'queryStringCaching' => 1,

        'includedQueryStringParams' => ['campid','id','fm','watchwebinar','elqTrack','r'],

        'excludedQueryStringParams' => ['\b(?!(campid|id|fm|watchwebinar|elqTrack|r)\b)\w+'],

        'warmPagesWithQueryStringParams' => false,

        'excludedUriPatterns' => [
           ***** hidden ******
        ],

        // The URI patterns to include in caching. Set `siteId` to a blank string to indicate all sites.
        'includedUriPatterns' => [
           [
               'siteId' => "", 'uriPattern' => '.*',
           ],
        ],

        'injectScriptEvent' => 'DOMContentLoaded',

        'integrations' => [
           'putyourlightson\blitz\drivers\integrations\FeedMeIntegration',
           'putyourlightson\blitz\drivers\integrations\SeomaticIntegration',
        ],

        'maxRetryAttempts' => 10,

        'mutexTimeout' => 1,

        'refreshCacheAutomaticallyForGlobals' => true,

        'refreshCacheJobPriority' => 10,

        'refreshCacheWhenElementSavedUnchanged' => false,

        'refreshCacheWhenElementSavedNotLive' => false,
```

**Comment by bencroker** [comment]
To get an idea of how many pages are cached and how many of those are related to entries, can you check how many rows are in the `blitz_caches` and `blitz_elementcaches` database tables?

> Clear cache via URL API just throw a timeout when we are full cached, all pages already warmed up

It’s more reliable to run this via the `php craft blitz/cache/clear` console command, which should not be limited by resources nor time, see https://putyourlightson.com/plugins/blitz#console-commands

> Accessing CMS > Utilities > Blitz just throw time out, it doesn't even load

Disabling `countCachedFiles` should help with this.

```php
        'cacheStorageSettings' => [
           'folderPath' => '@webroot/cache/blitz',
           'createGzipFiles' => false,
           'countCachedFiles' => false,
        ],
```

> Queue job for clearing cache are stuck or take more than 2 hours, even when is just for 1 entry

For _clearing_ the cache or _refreshing_ the cache?

> Scheduled posts never show up on the main listing, most likely due to the clear cache taking way too long

Do you have a cron job set up to refresh expired cache? See https://putyourlightson.com/plugins/blitz#cron-jobs

**Comment by brnquester** [comment]
@bencroker - thank you for the quick response. See my replies below:

> To get an idea of how many pages are cached and how many of those are related to entries, can you check how many rows are in the blitz_caches and blitz_elementcaches database tables?

`blitz_caches` = 14631
`blitz_elementcaches` = 886313

> For clearing the cache or refreshing the cache?

Refreshing. Sorry.

> Do you have a cron job set up to refresh expired cache?

No. We thought it would be added to the queue job automatically.

**Comment by bencroker** [comment]
So on average there are 60 entries per cached page, which seems like a lot. Does that sound accurate to you?

Since you have `cacheElementQueries` disabled, I would expect the `blitz_elementquerycaches` database table to be empty, is that the case?

**Comment by brnquester** [comment]
> So on average there are 60 entries per cached page, which seems like a lot. Does that sound accurate to you?

Just to clarify, what exactly does that entails? Entries from `craft_entires` / `craft_contents`? BTW - We use super table and Neo. Also, our set up is an upgrade from Craft v2.

> Since you have cacheElementQueries disabled, I would expect the blitz_elementquerycaches database table to be empty, is that the case?

Yes. That's the case. The table is empty.

**Comment by brnquester** [comment]
@bencroker - again, thank you very much! So here are my take aways of this thread:

1. Use `php craft blitz/cache/clear` instead of API calls due to the size of the cached storage
2. Change `countCachedFiles` to `false` to avoid the `CMS > Utilities > Blitz` hanging forever
3. Add cron job for `blitz/cache/refresh-expired` 

Any other suggestion? I'll be implementing all of that today!

**Comment by brnquester** [comment]
@bencroker - when setting `countCachedFiles` to `false` the CMS throws an error of "TypeError: Unsupported operand types: string - string" for the line below: 
https://github.com/putyourlightson/craft-blitz/blob/e6efa943a2ff87b3d1f9e2af41a9a1f3b55f81e1/src/drivers/storage/FileStorage.php#L226-L228 

It tries to decrement `-=` a string `-` which what the function of count return here:
https://github.com/putyourlightson/craft-blitz/blob/e6efa943a2ff87b3d1f9e2af41a9a1f3b55f81e1/src/drivers/storage/FileStorage.php#L338-L340 


Do you want me to open another issue as a bug report for this one? I'll patch it for now until there is a fix on the release for v3

**Comment by bencroker** [comment]
Thanks, fixed that bug in https://github.com/putyourlightson/craft-blitz/commit/24b0dce748f5614ece0c3c5ac55679853b5f154f and released in version 3.12.9.
